Manually update katsu's OpenAPI spec to v3.0 data model

OrdiNeu opened the pull request:
# Description
This implements the CanDIGv3 data model changes as outlined in [this document](https://docs.google.com/document/d/12GACkvTxKc0KhmV3bS8BIySu0qwPBfPzaKqp8NrXaaU/edit).

The updated schema can be found at [this link (includes only the diffs to the schema)](https://github.com/CanDIG/katsu/pull/239/files/119fee76a237de9cbb93330876c25a17486c0800..ed7311f059bf3296de477356eb2afe33c5240f90).

## Intended list of changes

Changes to the models includes:

Moving a few fields around: 

1. `primary_site` from donor to primary diagnosis, though the endpoints are still able to query based on it
2. `pathological_*`
3. `[“days_per_cycle”, “number_of_cycles”]`, 

Removing a few fields:

1. Surgery's `submitter_specimen_id`
2. Follow-up's `recurrence_*`
3. Treatment's `[“line_of_treatment”, “treatment_setting”]`
4. Primary Diagnosis' `[“lymph_nodes_examined_status”, “lymph_nodes_examined_method”, “number_lymph_nodes_positive”]`

Adding a few fields: 

1. Specimen's `submitter_treatment_id`
2. `[“surgery_reference_database”, “surgery_reference_identifier”]`

And adding the new Systemic Therapy table, which supercedes (and removes) Chemo/Hormonal/Immunotherapy.

This PR does not address how the implementation of the new table, nor the new/changed endpoints, but only outlines what will need to be changed

OrdiNeu committed:
Remove fields [“lymph_nodes_examined_status”, “lymph_nodes_examined_method”, “number_lymph_nodes_positive”]

OrdiNeu committed:
Surgery remove the optional field “submitter_specimen_id”

OrdiNeu committed:
Added required fields [“surgery_reference_database”, “surgery_reference_identifier”]

Assuming there's no 64 character limit on the
surgery_reference_identifier yet

OrdiNeu committed:
Remove follow-up optional fields

[“recurrence_tumour_staging_system”,
“recurrence_stage_group”, “recurrence_t_category”,
“recurrence_n_category”, “recurrence_m_category”]

OrdiNeu committed:
Add new Systemic Therapy table, remove old chemo/immuno/hormone therapy tables

OrdiNeu committed:
Add start/end dates to the systemic therapy schemas

OrdiNeu committed:
Cleanup: Make systemic_therapy_type required, remove submitter_treatment_id from the SampleRegistrations I accidentally added it to

OrdiNeu committed:
Copy over the schema to a new file

OrdiNeu committed:
Working on a new schema (written manually) to cover changes to v3 of the data model

This has the following work done so far:
1. Move required field “primary_site” from Donor to Primary Diagnosis
2. Add optional “submitter_treatment_id” if the specimen was resected
   during surgery
3. Moved the conditional fields [“pathological_tumour_staging_system,
   “pathological_stage_group”,
   “pathological_t_category”,  “pathological_n_category”,
   “pathological_m_category”] from Specimen to Primary Diagnosis

OrdiNeu committed:
Moved optional fields [“days_per_cycle”, “number_of_cycles”] to Systemic Therapy

OrdiNeu committed:
Remove line_of_treatment and treatment_setting

OrdiNeu committed:
Fixes from the data portal meeting

mshadbolt commented on chord_metadata_service/mohpackets/docs/schema_manual.yml:
```diff
@@ -4895,6 +4385,20 @@ components:
+    SystemicTherapyTypeEnum:
+      enum:
+      - Chemotherapy
+      - Hormone Therapy
```
Spreadsheet says this should have lowercase on the therapy
```suggestion 
      - Hormone therapy
```

mshadbolt commented on chord_metadata_service/mohpackets/docs/schema_manual.yml:
```diff
@@ -420,16 +460,17 @@ components:
       properties:
         actual_cumulative_drug_dose:
           anyOf:
           - type: integer
```
This might have been an issue not picked up before but the spreadsheet specifies this as 'number' rather than 'integer' so not sure we can assume non decimal numbers
```suggestion
          - type: number
```

mshadbolt commented on chord_metadata_service/mohpackets/docs/schema_manual.yml:
```diff
@@ -454,6 +495,31 @@ components:
           - type: integer
```
Same comment as above
```suggestion
          - type: number
```

mshadbolt commented on chord_metadata_service/mohpackets/docs/schema_manual.yml:
```diff
@@ -1964,14 +1627,14 @@ components:
       properties:
         actual_cumulative_drug_dose:
           anyOf:
           - type: integer
```
```suggestion
          - type: number
```

mshadbolt commented on chord_metadata_service/mohpackets/docs/schema_manual.yml:
```diff
@@ -1996,7 +1659,32 @@ components:
           - type: integer
```
```suggestion
          - type: number
```

mshadbolt commented on chord_metadata_service/mohpackets/docs/schema_manual.yml:
```diff
@@ -2113,80 +1781,33 @@ components:
-          - type: 'null'
-          title: Drug Name
-        drug_reference_database:
+        pathological_m_category:
```
I think this should be an enum

mshadbolt commented on chord_metadata_service/mohpackets/docs/schema_manual.yml:
```diff
@@ -2113,80 +1781,33 @@ components:
           - type: 'null'
-        drug_reference_identifier:
+          title: Pathological M Category
+        pathological_n_category:
```
I think this should have an enum ref

mshadbolt commented on chord_metadata_service/mohpackets/docs/schema_manual.yml:
```diff
@@ -2113,80 +1781,33 @@ components:
-          title: Drug Reference Identifier
-        immunotherapy_drug_dose_units:
+          title: Pathological N Category
+        pathological_stage_group:
```
I think this should have an enum ref

mshadbolt commented on chord_metadata_service/mohpackets/docs/schema_manual.yml:
```diff
@@ -2113,80 +1781,33 @@ components:
           - type: 'null'
-        immunotherapy_type:
+          title: Pathological Stage Group
+        pathological_t_category:
```
I think this should have an enum ref

mshadbolt commented on chord_metadata_service/mohpackets/docs/schema_manual.yml:
```diff
@@ -2113,80 +1781,33 @@ components:
           - type: 'null'
-        prescribed_cumulative_drug_dose:
+          title: Pathological T Category
+        pathological_tumour_staging_system:
```
this should have an enum ref

mshadbolt reviewed (changes requested):
Found a few small things but overall looks really good to me! Thanks!

OrdiNeu committed:
Apply suggestions by @mshadbolt

- Switch pathological categories to be an enum
- Switch actual/perscribed_cumulative_drug_dose to numbers

mshadbolt reviewed (approved):
LGTM

SonQBChau commented:
- Line 1044: remove `primary_site` object.
Otherwise, it looks good to me.

OrdiNeu committed:
Remove primary site from the DonorWithClinicalData schema?

kcranston commented:
> * Line 1044: remove `primary_site` object.
>   Otherwise, it looks good to me.

Confused about why we are removing `primary_site`. It still exists under Primary Diagnosis in the new model.

OrdiNeu commented:
> > * Line 1044: remove `primary_site` object.
> >   Otherwise, it looks good to me.
> 
> Confused about why we are removing `primary_site`. It still exists under Primary Diagnosis in the new model.

It'll be in the DonorWithClinicalData response under the NestedPrimaryDiagnosis schema, so to get the `primate_site` you'd go: `DonorWithClinicalData->primary_diagnoses[i]->primary_site`.

kcranston reviewed (changes requested):
Couple of documentation requests (to help both internal and external folks looking for the updated schema):
* can we change the name of the file to include `v3.0`?
* update the README to point to the new file

OrdiNeu committed:
Move schema_v3 -> schema_manual and add stuff to the README?

kcranston commented:
Sorry, that last comment about the readme wasn't clear. I don't think we want to replace all of the instances of `schema` with `schema_v3` in the readme until we have updated katsu. I just wanted a line in the readme that mentioned the new, manually created schema for the upcoming model change. I know you are away next week - we can fix this in your absence!

daisieh reviewed (approved):
LGTM

mshadbolt committed:
revert and add note

mshadbolt committed:
typo

mshadbolt commented on chord_metadata_service/mohpackets/docs/README.MD:
```diff
@@ -4,7 +4,10 @@ This folder contains the schema and documentation for **MoH models**
+To view the API documentation, simply open [schema.md](schema.md) or [Redoc](https://redocly.github.io/redoc/?url=https://raw.githubusercontent.com/CanDIG/katsu/feature/manual_schema/chord_metadata_service/mohpackets/docs/schema_manual.yml).
+
+> [!IMPORTANT]
+> Katsu is currently being transitioned to MoH clinical data model v3.0.0. There is a manually created v3.0.0 of the OpenAPI schema at [schema_v3.yml](schema_v3.yml) that can be used as a reference in the interim period. This file will be replaced by the automatically generated schema documentaion files once katsu is fully transitioned to model v3.0.0.
```
@kcranston is this what you were thinking for the line in the README?
